packages/arb-token-bridge-ui/src/token-bridge-sdk/CctpTransferStarter.ts

```diff
@@ -0,0 +1,140 @@
+import { fetchErc20Allowance } from '../util/TokenUtils'
+import { TokenMessengerAbi } from '../util/cctp/TokenMessengerAbi'
+
+export class CctpTransferStarter extends BridgeTransferStarter {
+  public transferType: TransferType
+
+  constructor(props: BridgeTransferStarterProps) {
+    super(props)
+    this.transferType = 'cctp'
+  }
+
+  public requiresNativeCurrencyApproval = async () => false
+
+  public approveNativeCurrency = async () => {
+    return
+  }
+
+  public async requiresTokenApproval({
+    amount,
+    signer,
+    destinationAddress
+  }: RequiresTokenApprovalProps): Promise<boolean> {
+    const sourceChainId = await getChainIdFromProvider(this.sourceChainProvider)
+
+    const address = await getAddressFromSigner(signer)
+
+    const { usdcContractAddress } = getCctpContracts({ sourceChainId })
+
+    const recipient = destinationAddress ?? address
+    const { tokenMessengerContractAddress } = getCctpContracts({
+      sourceChainId
+    })
+
+    const allowance = await fetchErc20Allowance({
+      address: usdcContractAddress,
+      provider: this.sourceChainProvider,
+      owner: recipient,
+      spender: tokenMessengerContractAddress
+    })
+
+    return allowance.lt(amount)
+  }
+
+  public async approveToken({ signer, amount }: ApproveTokenProps) {
+    const sourceChainId = await getChainIdFromProvider(this.sourceChainProvider)
+
+    const { usdcContractAddress, tokenMessengerContractAddress } =
+      getCctpContracts({ sourceChainId })
+
+    // approve USDC token for burn
+    const contract = ERC20__factory.connect(usdcContractAddress, signer)
+    const tx = await contract.functions.approve(
+      tokenMessengerContractAddress,
+      amount ?? constants.MaxInt256
+    )
+    return tx
+  }
+
+  public async approveTokenEstimateGas({ signer, amount }: ApproveTokenProps) {
+    const sourceChainId = await getChainIdFromProvider(this.sourceChainProvider)
+    const { usdcContractAddress, tokenMessengerContractAddress } =
+      getCctpContracts({ sourceChainId })
+    const contract = ERC20__factory.connect(usdcContractAddress, signer)
+    return await contract.estimateGas.approve(
+      tokenMessengerContractAddress,
+      amount ?? constants.MaxInt256
+    )
+  }
+
+  public async transfer({ signer, amount, destinationAddress }: TransferProps) {
+    const sourceChainId = await getChainIdFromProvider(this.sourceChainProvider)
+
+    const address = await getAddressFromSigner(signer)
+
+    // cctp has an upper limit for transfer
+    const burnLimit = await fetchPerMessageBurnLimit({
+      sourceChainId
+    })
+
+    if (burnLimit.lte(amount)) {
+      const formatedLimit = formatAmount(burnLimit, {
+        decimals: 6, // hardcode for USDC
+        symbol: 'USDC'
+      })
+      throw Error(
+        `The limit for transfers using CCTP is ${formatedLimit}. Please lower your amount and try again.`
+      )
+    }
+
+    const recipient = destinationAddress || address
```

Comment:
I see in a few spots we used ?? and now we use ||, we should be consistent

Reply:
Yeah `destinationAddress` can be string | undefined, so `??` would play nicely here. || operator is also being used in one more place apart from this, but I'm not touching that to keep the scope of this PR limited.